tao_compiler/mlir/disc/transforms/dot_rewriter.cc

```diff
@@ -100,6 +105,195 @@ struct DotToDotGeneralConvert : public OpRewritePattern<mhlo::DotOp> {
         op.getLoc(), op.getType(), op.lhs(), op.rhs(), dot_dimension_attr,
         nullptr);
     rewriter.replaceOp(op, dot_general);
+    return success();
+  }
+};
+
+SmallVector<Value, 4> getDimSizesOfTensor(PatternRewriter& rewriter,
+                                          Operation* op, Value value) {
+  auto value_ty = value.getType().dyn_cast<RankedTensorType>();
+
+  auto loc = op->getLoc();
+  auto rank = value_ty.getRank();
+  // Get int vector [0, 1, ..., rank-1]
+  SmallVector<Value, 4> dim_sizes;
+  for (size_t d = 0; d < rank; ++d) {
+    dim_sizes.emplace_back(rewriter.create<tensor::DimOp>(loc, value, d));
+  }
+  return dim_sizes;
+}
+
+Value unsqueezeTensorDim(PatternRewriter& rewriter, Operation* op, Value tensor,
+                         int64_t dim) {
+  // Returns a new tensor with dims of size 1 inserted at the specified
+  // position.
+  //
+  // The position indices (must be high to low dimension number of the returned
+  // tensor) are specified with unsqzDims. Indices must be in-order, and in
+  // range of tensor rank. Thus, unsqueeze a rank 1 tensor with {0, 2}, {0, 1,
+  // 3}, {0, 1, 2} are all valid dimension sets, but {0, 3}, {2} are not.
+  auto dim_sizes = getDimSizesOfTensor(rewriter, op, tensor);
+  auto loc = op->getLoc();
+  auto one = rewriter.create<arith::ConstantIndexOp>(loc, 1);
+  auto rank = dim_sizes.size();
+
+  auto rankTy = tensor.getType().dyn_cast<RankedTensorType>();
+  auto oldShape = rankTy.getShape();
+
+  SmallVector<Value, 4> newDimSizes;
+  std::vector<int64_t> newShape;
```

Comment:
nit: just curious why `newShape` is not a `SmallVector` but `newDimSizes` is. :P